File: packages/java/anip-example-flights-quarkus/src/main/java/dev/anip/example/flights/BookFlightCapability.java
package dev.anip.example.flights;

import dev.anip.core.ANIPError;
import dev.anip.core.CapabilityDeclaration;
import dev.anip.core.CapabilityInput;
import dev.anip.core.CapabilityOutput;
import dev.anip.core.CapabilityRequirement;
import dev.anip.core.Constants;
import dev.anip.core.Cost;
import dev.anip.core.CostActual;
import dev.anip.core.SideEffect;
import dev.anip.service.CapabilityDef;
import dev.anip.service.InvocationContext;

import java.security.SecureRandom;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Book a confirmed flight reservation.
 */
public final class BookFlightCapability {

    private BookFlightCapability() {}

    private static final SecureRandom RANDOM = new SecureRandom();

    public static CapabilityDef create() {
        CapabilityDeclaration decl = new CapabilityDeclaration(
                "book_flight",
                "Book a confirmed flight reservation",
                "1.0",
                List.of(
                        new CapabilityInput("flight_number", "string", true, "Flight to book"),
                        new CapabilityInput("date", "date", true, "Travel date (YYYY-MM-DD)"),
                        new CapabilityInput("passengers", "integer", false, 1, "Number of passengers")
                ),
                new CapabilityOutput("booking_confirmation",
                        List.of("booking_id", "flight_number", "departure_time", "total_cost")),
                new SideEffect("irreversible", "none"),
                List.of("travel.book"),
                new Cost(
                        "estimated",
                        Map.of(
                                "currency", "USD",
                                "estimated_range", Map.of("min", 280, "max", 500)
                        ),
                        "search_flights",
                        null, null, null
                ),
                List.of(new CapabilityRequirement("search_flights",
                        "must select from available flights before booking")),
                List.of("unary")
        );

        return new CapabilityDef(decl, BookFlightCapability::handle);
    }

    private static Map<String, Object> handle(InvocationContext ctx, Map<String, Object> params) {
        String flightNumber = (String) params.get("flight_number");
        String date = (String) params.get("date");
        int passengers = 1;
        if (params.get("passengers") instanceof Number p) {
            passengers = p.intValue();
        }

        if (flightNumber == null || flightNumber.isEmpty()
                || date == null || date.isEmpty()) {
            throw new ANIPError(Constants.FAILURE_INVALID_PARAMETERS,
                    "flight_number and date are required");
        }

        Map<String, Object> flight = FlightData.findFlight(flightNumber, date);
        if (flight == null) {
            throw new ANIPError(Constants.FAILURE_UNAVAILABLE,
                    "flight " + flightNumber + " on " + date + " not found");
        }

        double price = ((Number) flight.get("price")).doubleValue();
        double totalCost = price * passengers;
        String currency = (String) flight.get("currency");

        // Generate booking ID.
        byte[] b = new byte[4];
        RANDOM.nextBytes(b);
        String bookingId = String.format("BK-%02X%02X%02X%02X", b[0], b[1], b[2], b[3]);

        // Track actual cost.
        ctx.setCostActual(new CostActual(
                Map.of("amount", totalCost, "currency", currency),
                null
        ));

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("booking_id", bookingId);
        result.put("flight_number", flight.get("flight_number"));
        result.put("departure_time", flight.get("departure_time"));
        result.put("total_cost", totalCost);
        result.put("currency", currency);
        return result;
    }
}

File: packages/csharp/src/Anip.Example.Flights/BookFlightCapability.cs
using System.Security.Cryptography;
using Anip.Core;
using Anip.Service;

namespace Anip.Example.Flights;

/// <summary>
/// Book a confirmed flight reservation.
/// </summary>
public static class BookFlightCapability
{
    public static CapabilityDef Create()
    {
        var declaration = new CapabilityDeclaration
        {
            Name = "book_flight",
            Description = "Book a confirmed flight reservation",
            ContractVersion = "1.0",
            Inputs = new List<CapabilityInput>
            {
                new() { Name = "flight_number", Type = "string", Required = true, Description = "Flight to book" },
                new() { Name = "date", Type = "date", Required = true, Description = "Travel date (YYYY-MM-DD)" },
                new() { Name = "passengers", Type = "integer", Required = false, Default = 1, Description = "Number of passengers" },
                new() { Name = "quote_id", Type = "object", Required = true, Description = "Priced quote from search_flights" },
            },
            Output = new CapabilityOutput
            {
                Type = "booking_confirmation",
                Fields = new List<string> { "booking_id", "flight_number", "departure_time", "total_cost" },
            },
            SideEffect = new SideEffect
            {
                Type = "irreversible",
                RollbackWindow = "none",
            },
            MinimumScope = new List<string> { "travel.book" },
            Cost = new Cost
            {
                Certainty = "estimated",
                Financial = new FinancialCost
                {
                    Currency = "USD",
                    RangeMin = 280,
                    RangeMax = 500,
                },
                DeterminedBy = "search_flights",
            },
            Requires = new List<CapabilityRequirement>
            {
                new()
                {
                    Capability = "search_flights",
                    Reason = "must select from available flights before booking",
                },
            },
            RequiresBinding = new List<BindingRequirement>
            {
                new()
                {
                    Type = "quote",
                    Field = "quote_id",
                    SourceCapability = "search_flights",
                    MaxAge = "PT15M",
                },
            },
            ControlRequirements = new List<ControlRequirement>(),
            RefreshVia = new List<string> { "search_flights" },
            VerifyVia = new List<string>(),
            ResponseModes = new List<string> { "unary" },
        };

        return new CapabilityDef(declaration, Handle);
    }

    private static Dictionary<string, object?> Handle(InvocationContext ctx, Dictionary<string, object?> parameters)
    {
        var flightNumber = GetString(parameters, "flight_number");
        var date = GetString(parameters, "date");
        var passengers = GetInt(parameters, "passengers", 1);
        parameters.TryGetValue("quote_id", out var quoteId);

        if (string.IsNullOrEmpty(flightNumber) || string.IsNullOrEmpty(date))
        {
            throw new AnipError(Constants.FailureInvalidParameters,
                "flight_number and date are required");
        }

        var flight = FlightData.FindFlight(flightNumber, date);
        if (flight == null)
        {
            throw new AnipError(Constants.FailureUnavailable,
                $"flight {flightNumber} on {date} not found");
        }

        var price = Convert.ToDouble(flight["price"]);
        var totalCost = price * passengers;
        var currency = (string)flight["currency"]!;

        // Generate booking ID.
        var bytes = new byte[4];
        RandomNumberGenerator.Fill(bytes);
        var bookingId = $"BK-{Convert.ToHexString(bytes)}";

        // Track actual cost.
        ctx.CostActual = new CostActual
        {
            Financial = new FinancialCost
            {
                Amount = totalCost,
                Currency = currency,
            },
        };

        return new Dictionary<string, object?>
        {
            ["booking_id"] = bookingId,
            ["flight_number"] = flight["flight_number"],
            ["departure_time"] = flight["departure_time"],
            ["total_cost"] = totalCost,
            ["currency"] = currency,
        };
    }

    private static string? GetString(Dictionary<string, object?> parameters, string key)
    {
        if (!parameters.TryGetValue(key, out var value))
            return null;

        if (value is System.Text.Json.JsonElement el &&
            el.ValueKind == System.Text.Json.JsonValueKind.String)
        {
            return el.GetString();
        }

        return value as string;
    }

    private static int GetInt(Dictionary<string, object?> parameters, string key, int defaultValue)
    {
        if (!parameters.TryGetValue(key, out var value))
            return defaultValue;

        if (value is System.Text.Json.JsonElement el)
        {
            if (el.ValueKind == System.Text.Json.JsonValueKind.Number && el.TryGetInt32(out var intVal))
                return intVal;
        }

        if (value is int i) return i;
        if (value is long l) return (int)l;
        if (value is double d) return (int)d;

        return defaultValue;
    }
}
